enlighten/Controller.py

```diff
@@ -586,24 +586,34 @@ def check_ready_initialize(self):
         in_progress_specs = list(self.multispec.in_process.items())
         for device_id, device in in_progress_specs:
             log.debug(f"checking {device_id} for settings results")
-            if device.poll_settings():
-                self.header("connect_new: successfully connected %s" % device_id)
-                self.initialize_new_device(device)
-
-                # remove the "in-process" flag, as it's now in the "connected" list
-                # and fully initialized
-                if device.is_ble:
-                    self.form.ui.readingProgressBar.show()
-                self.multispec.remove_in_process(device_id)
-                self.header("connect_new: done (%s)" % device_id)
+            disconnect_device = False
+
+            poll_result = device.poll_settings()
+            if poll_result is not None:
+                if poll_result.data:
+                    self.header("connect_new: successfully connected %s" % device_id)
+                    self.initialize_new_device(device)
+
+                    # remove the "in-process" flag, as it's now in the "connected" list
+                    # and fully initialized
+                    if device.is_ble:
+                        self.form.ui.readingProgressBar.show()
+                    self.multispec.remove_in_process(device_id)
+                    self.header("connect_new: done (%s)" % device_id)
+                else:
+                    self.marquee.error(poll_result.error_msg)
+                    disconnect_device = True
+                    
             else:
                 # didn't get settings so check for timeout
                 if device.connect_start_time + datetime.timedelta(seconds=self.spec_timeout) < datetime.datetime.now():
                     log.error(f"{device_id} settings timed out, giving up on the spec")
-                    device.disconnect()
-                    self.multispec.set_gave_up(device_id)
-                    self.multispec.remove_in_process(device_id)
+                    disconnect_device = True
 
+            if disconnect_device:
```

Comment:
I'm fine with this. Trying to understand though, what is the purpose of moving it outside of the if else and using a flag?

Reply:
To avoid duplicate code.  There's now two places in the above logic which might trigger a disconnect sequence (either a False SpectrometerResponse or a None+timeout).